File: OC/DCUtils/DCUtils.m
//
//  DCUtils.m
//  DCModel
//
//  Created by Carden on 2018/1/18.
//  Copyright © 2018年 Carden. All rights reserved.
//

#import "DCUtils.h"

@implementation DCUtils
#pragma mark - bundleVersion
+ (NSString *)dc_bundleVersion {

    return [[[NSBundle mainBundle]infoDictionary]valueForKey:@"CFBundleShortVersionString"];
}
#pragma mark - AppleStore
+ (void)dc_jumpToAppleStore {

    NSString *iTunesLink = [NSString stringWithFormat:@"itms-apps://itunes.apple.com/cn/app/id%@?mt=8",@"1043010693"];
    
    if (@available(iOS 10.0, *)) {
        [[UIApplication sharedApplication]openURL:[NSURL URLWithString:iTunesLink] options:@{UIApplicationOpenURLOptionUniversalLinksOnly : @YES} completionHandler:nil];
    } else {
       [[UIApplication sharedApplication] openURL:[NSURL URLWithString:iTunesLink]];
    }


}
#pragma mark - UserDefault
+ (void)dc_saveData:(NSDictionary *)data {

    [data enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        
        if (obj && ![obj isEqual:[NSNull null]]) {
            [self dc_setObject:[NSString stringWithFormat:@"%@",obj] key:key];
        }
    }];
}
+ (void)dc_setObject:(id)data key:(NSString *)key {

    if (data) {
        [[NSUserDefaults standardUserDefaults] setObject:data forKey:key];
        [[NSUserDefaults standardUserDefaults] synchronize];
    }
}

+ (id)dc_getObjectForKey:(NSString *)key {

    return  [[NSUserDefaults standardUserDefaults] objectForKey:key];
}
+ (void)dc_setBool:(BOOL)data key:(NSString *)key {

    if (data) {
    [[NSUserDefaults standardUserDefaults]setBool:data forKey:key];
    [[NSUserDefaults standardUserDefaults] synchronize];
    }
}

+ (BOOL)dc_getBoolForKey:(NSString *)key {

    return  [[NSUserDefaults standardUserDefaults] boolForKey:key];
}
+ (void)dc_deleteObject:(NSString *)key
{
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:key];
    [[NSUserDefaults standardUserDefaults] synchronize];
}
#pragma mark - Cookies
+ (void)dc_saveCookies {
    
    NSData *cookiesData = [NSKeyedArchiver archivedDataWithRootObject: [[NSHTTPCookieStorage sharedHTTPCookieStorage] cookies]];
    [self dc_setObject:cookiesData key:@"DCCookies"];
}


+ (void)dc_loadCookies {

    NSArray *cookies = [NSKeyedUnarchiver unarchiveObjectWithData: [self dc_getObjectForKey:@"DCCookies"]];
    NSHTTPCookieStorage *cookieStorage = [NSHTTPCookieStorage sharedHTTPCookieStorage];
    for (NSHTTPCookie *cookie in cookies){
        [cookieStorage setCookie: cookie];
    }
}
#pragma mark - callPhone
+ (void)dc_callNumber:(NSString *)phone {

    NSString *telUrl = [NSString stringWithFormat:@"telprompt:%@",phone];
    if (@available(iOS 10.0, *)) {
        [[UIApplication sharedApplication]openURL:[NSURL URLWithString:telUrl] options:@{UIApplicationOpenURLOptionUniversalLinksOnly : @YES} completionHandler:nil];
    } else {
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:telUrl]];
    }

}
@end

File: Swift/DCUtils/DCUtils.swift
//
//  DCUtils.swift
//  分类
//
//  Created by Carden on 2018/2/7.
//  Copyright © 2018年 Carden. All rights reserved.
//

import UIKit
class DCUtils: NSObject {
    //判断版本号
    class  func dc_isNeedUpdate(ver:String) -> Bool {
        let preVer = dc_bundleVersion()
        let curVer = ver
        //比较两个版本
        let preArr = preVer.split(separator: ".")
        let curArr = curVer.split(separator: ".")
        let pV:Int = Int(preArr.joined(separator: "")) ?? 0
        let cV:Int = Int(curArr.joined(separator: "")) ?? 0
        return cV > pV
    }
    //获取版本号
    class  func dc_bundleVersion() -> String {
        
        return Bundle.main.infoDictionary!["CFBundleShortVersionString"] as! String
    }
    //跳转AppStore
    class func dc_jumpToAppleStore() {
        let string = "itms-apps://itunes.apple.com/cn/app/id1043010693?mt=8"
        if #available(iOS 10.0, *) {
            UIApplication.shared.open(URL.init(string: string)!, options: [UIApplication.OpenExternalURLOptionsKey.universalLinksOnly : true], completionHandler: nil)
         }else {
            UIApplication.shared.openURL(URL.init(string: string)!)
        }
    }
    ///打电话
    class func dc_callNumber(phone:String) {
        let telUrl = "telprompt:\(phone)"
        if #available(iOS 10.0, *) {
            UIApplication.shared.open(URL.init(string: telUrl)!, options: [UIApplication.OpenExternalURLOptionsKey.universalLinksOnly : true], completionHandler: nil)
        }else {
            UIApplication.shared.openURL(URL.init(string: telUrl)!)
        }
        
    }
    /// 验证手机号
    class func dc_verifyPhoneNumber(phoneNumber:String) -> Bool {
        let  MOBILE = "^1[0-9]{10}$";
        let regextestmobile = NSPredicate(format: "SELF MATCHES %@",MOBILE)
        return  regextestmobile.evaluate(with: phoneNumber)
    }
    /// 验证密码
    class func dc_verifyPassword(password:String) -> Bool {
        let  pwd = "^(?![0-9]+$)(?![a-zA-Z]+$)[a-zA-Z0-9]{8,16}"
        let regextestmobile = NSPredicate(format: "SELF MATCHES %@",pwd)
        return  regextestmobile.evaluate(with: password)
    }
    /// 验证验证码
    class func dc_verifyVerCode(code:String) -> Bool {
        let pwd  = "^[0-9]{6}$"
        let regextestmobile = NSPredicate(format: "SELF MATCHES %@",pwd)
        return  regextestmobile.evaluate(with: code)
    }
    /// 验证身份证号
    class func dc_verifyID(ID:String) -> Bool {
        let pwd = "^(\\d{14}|\\d{17})(\\d|[xX])$"
        let regextestmobile = NSPredicate(format: "SELF MATCHES %@",pwd)
        return  regextestmobile.evaluate(with: ID)
    }
    ///返回密文手机号
    class func dc_hidePhone(phoneNumber:String) -> String {
        if self.dc_verifyPhoneNumber(phoneNumber: phoneNumber) {
            let phone = NSString.init(string: phoneNumber)
            return phone.replacingCharacters(in: NSMakeRange(3, 5), with:"*****")
        }
        return "*****";
    }
    ///返回对应周几
    class func dc_weekDay(index:Int) -> String {
        var week = "未知";
        switch (index) {
        case 0:
            week = "周日";
            break;
        case 1:
            week =  "周一";
            break;
        case 2:
            week =  "周二";
            break;
        case 3:
            week =  "周三";
            break;
        case 4:
            week =  "周四";
            break;
        case 5:
            week =  "周五";
            break;
        case 6:
            week =  "周六";
            break;
        default:
            break;
        }
        return week
    }
    //MARK:- UserDefault
    //保存 字典
    class func dc_saveData(data:[String:AnyObject]) {
        for (key,value) in data {
            dc_setObject(data: value, key: key)
        }
    }
    class  func dc_setObject(data:Any,key:String) {
       
        UserDefaults.standard.set(data, forKey: key)
        
    }
    class   func dc_getObjectForKey(key:String)->Any? {
        
        return  UserDefaults.standard.object(forKey: key)
    }
    class   func dc_deleteObject(key:String){
        return  UserDefaults.standard.removeObject(forKey: key)
    }
    // MARK: - cookies
    class func dc_saveCookies() {
        var cookiesData:Data;
        if #available(iOS 12.0, *) {
            cookiesData = try! NSKeyedArchiver.archivedData(withRootObject: HTTPCookieStorage.shared.cookies as Any, requiringSecureCoding: false)
        }else {
            cookiesData = NSKeyedArchiver.archivedData(withRootObject: HTTPCookieStorage.shared.cookies as Any)
            
        }
        self.dc_setObject(data: cookiesData, key: "DCCookies")
    }
    class func dc_loadCookies() {
        if (dc_getObjectForKey(key: "DCCookies") == nil) {
            return
        }
        let  cookies = NSKeyedUnarchiver.unarchiveObject(with: self.dc_getObjectForKey(key: "DCCookies") as! Data) as! [HTTPCookie]
        let cookieStorage = HTTPCookieStorage.shared
        for cookie in cookies {
            cookieStorage.setCookie(cookie)
        }
        
    }
    class func dc_deleteCookies() {
        if (dc_getObjectForKey(key: "DCCookies") == nil) {
            return
        }
        let cookieStorage = HTTPCookieStorage.shared
        for cookie in  HTTPCookieStorage.shared.cookies! {
            cookieStorage.deleteCookie(cookie)
        }
        self.dc_deleteObject(key: "DCCookies")
    }
}
